### PR TITLE
Added full-screen modal CKEditor on mobile web (closes #538)

### LINKED ISSUE #538: Open CKEditor in full-screen on mobile web
It's can be difficult to enter content in the Moments/Strategy form on mobile web because the on-screen keyboard covers most of the page.

Desired behaviour:
1) User clicks the CKEditor
2) Full-screen modal opens with just the CKEditor for that input field
3) User types in content
4) When they hit the "x" icon, the text they wrote to the editor appears in form

image

### PR DESCRIPTION
**Pull request by alvarocasadoc**
Current behaviour:
- Only when the **viewport** is small enough (**< 640px**).
- User taps one of the ckeditors and a new modal is displayed with a ckeditor inside. This modal is implemented as a **new partial** (`views/shared/_editor_modal`) and it is included in `views/moments/_form` and `views/strategies/_form`.
- While this modal is visible, **page scrolling is disabled** to improve usability by leaving only the ckeditor field's scroll active.
- This editor has the **same contents** from the selected ckeditor.
- When the user taps the closing cross icon the modal disappears and the content appears in the previously selected ckeditor and **page scrolling is re-enabled**.
- Form is submitted and its contents are saved.

This behaviour is coded in `assets/javascripts/mobile_cke.js `and the styling is inside `assets/stylesheets/editor_modal.scss`.

demo

### CODE REVIEW

File: app/assets/javascripts/mobile_cke.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+var onReadyMobileCke = function() {
+  var MEDIUM_BREAKPOINT = 640;
+
+  var ui = null;
```
Comment (julianguyen) [inline app/assets/javascripts/mobile_cke.js]: You don't need to initialize this as null.

File: app/assets/javascripts/mobile_cke.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  var focused_editor = null;
```
Comment (julianguyen) [inline app/assets/javascripts/mobile_cke.js]: You don't need to initialize this as null too.

File: app/assets/javascripts/mobile_cke.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  var modal_editor = CKEDITOR.instances.modal_editor;
```
Comment (julianguyen) [inline app/assets/javascripts/mobile_cke.js]: You might want to do this `var modal_editor = CKEDITOR && CKEDITOR.instances && CKEDITOR.instances.modal_editor;` to make sure CKEDITOR is loaded onto the page.

File: app/assets/javascripts/mobile_cke.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+var onReadyMobileCke = function() {
```
Comment (julianguyen) [inline app/assets/javascripts/mobile_cke.js]: You should be testing this file through Jasmine.

File: app/assets/javascripts/mobile_cke.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    $('#editor_modal').removeClass('display_block');
+    $('#editor_modal').addClass('display_none');
+    focused_editor.setData(modal_editor.getData());
+    modal_editor.focus();
+    $('body').removeClass('no_scroll');
+  });
+
+  for (instance in CKEDITOR.instances) {
+    var editor = CKEDITOR.instances[instance];
+    if (editor.name != 'modal_editor') {
```
Comment (julianguyen) [inline app/assets/javascripts/mobile_cke.js]: Use `!==` and double quotes please :)

File: app/assets/javascripts/mobile_cke.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      editor.on('focus', function(event) {
+        if ($(window).width() <= MEDIUM_BREAKPOINT) {
+          $('body').addClass('no_scroll');
+          showEditorModal();
+          focused_editor = this;
```
Comment (julianguyen) [inline app/assets/javascripts/mobile_cke.js]: Is this guaranteed to be the correct `this`?

File: app/assets/javascripts/mobile_cke.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+          ui = jQuery(this.container.$);
```
Comment (julianguyen) [inline app/assets/javascripts/mobile_cke.js]: Can't you just use `$(this)` here?

File: app/assets/stylesheets/application/editor_modal.scss [hunk 8]
```diff
@@ -0,0 +1,74 @@
+#editor_modal {
+  width: 100%;
+  height: 100vh; //fill all screen height
```
Comment (julianguyen) [inline app/assets/stylesheets/application/editor_modal.scss]: Why doesn't putting 100% just work?

File: app/assets/javascripts/mobile_cke.js (hunk 4 above)
Comment (julianguyen) [inline app/assets/javascripts/mobile_cke.js]: I would rename this to mobile_editor.js

File: spec/javascripts/ModalEditor.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+
+describe("ModalEditor", function() {
```
Comment (julianguyen) [inline spec/javascripts/ModalEditor.js]: Please rename this file to `ModalEditorSpec.js`

File: spec/javascripts/ModalEditor.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  });
+
+  it("close modal", function() {
+    $("#close_editor").click();
+    expect($("#editor_modal").hasClass("display_none")).toBe(true);
+  });
+
+  it("keep modal open on modal click", function() {
+    $("#editor_modal").click();
+    expect($("#editor_modal").hasClass("display_none")).toBe(false);
```
Comment (julianguyen) [inline spec/javascripts/ModalEditor.js]: I would also test the appearance of #modal_text

**Comment by alvarocasadoc** [comment]
I'll review all your comments and write the required tests 😄

**Comment by alvarocasadoc** [comment]
I've made the changes you commented and wrote the feature test for the modal editor. I can run jasmine and all tests pass, but when I use `jasmine:ci` my tests continuously fail:
`ReferenceError: Can't find variable: onReadyModalEditor in http://localhost:46739/__spec__/ModalEditorSpec.js (line 4)`

Right now I'm figuring out what causes this and when it's fixed I'll commit the changes :)

File: spec/javascripts/ModalEditorSpec.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    $("#close_editor").click();
+    expect($("#editor_modal")).toBeHidden();
+    expect($("#modal_text")).toBeHidden();
+  });
+
+  it("should be kept open on modal click", function() {
+      showEditorModal();
+      $("#editor_modal").click();
+      expect($("#editor_modal")).toBeVisible();
+      expect($("#editor_modal")).toBeVisible();
```
Comment (julianguyen) [inline spec/javascripts/ModalEditorSpec.js]: Looks like this is a duplicate of the previous line.

File: spec/javascripts/ModalEditorSpec.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
+describe("ModalEditor", function() {
```
Comment (julianguyen) [inline spec/javascripts/ModalEditorSpec.js]: Regarding the issue you encountered with the tests failing on jasmine:ci, maybe resetting the modal after each it block is run will fix it? But you do have a beforeEach already 🤔

**Comment by alvarocasadoc** [comment]
I discovered that what was causing the issue was the same error ESLint was pointing at me. Now all checks pass correctly!

**Pull request merged**